File: backend/app/routes/journal.py
import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.auth import MessageResponse
from app.schemas.journal import (
    SessionCreate,
    SessionListResponse,
    SessionResponse,
    SessionUpdate,
)
from app.services import journal as journal_service

router = APIRouter(prefix="/journal/sessions", tags=["journal"])


@router.post("", response_model=SessionResponse, status_code=201)
async def create_session(
    body: SessionCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    session = await journal_service.create_session(
        db, current_user.id, body.model_dump(exclude_unset=True),
    )
    return SessionResponse.model_validate(session)


@router.get("", response_model=SessionListResponse)
async def list_sessions(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    offset: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    date_from: date | None = Query(None),
    date_to: date | None = Query(None),
    session_type: str | None = Query(None),
):
    items, total = await journal_service.list_sessions(
        db, current_user.id,
        offset=offset, limit=limit,
        date_from=date_from, date_to=date_to,
        session_type=session_type,
    )
    return SessionListResponse(
        items=[SessionResponse.model_validate(s) for s in items],
        total=total,
        offset=offset,
        limit=limit,
    )


@router.get("/{session_id}", response_model=SessionResponse)
async def get_session(
    session_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    session = await journal_service.get_session(db, current_user.id, session_id)
    return SessionResponse.model_validate(session)


@router.patch("/{session_id}", response_model=SessionResponse)
async def update_session(
    session_id: uuid.UUID,
    body: SessionUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    session = await journal_service.update_session(
        db, current_user.id, session_id, body.model_dump(exclude_unset=True),
    )
    return SessionResponse.model_validate(session)


@router.delete("/{session_id}", response_model=MessageResponse)
async def delete_session(
    session_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await journal_service.delete_session(db, current_user.id, session_id)
    return MessageResponse(message="Session deleted")

File: frontend/src/services/journal.ts
import apiClient from './api-client'
import type { SessionCreate, SessionUpdate, SessionResponse, SessionListResponse } from '../types'

export interface SessionFilters {
  offset?: number
  limit?: number
  date_from?: string
  date_to?: string
  session_type?: string
}

export async function getSessions(filters?: SessionFilters): Promise<SessionListResponse> {
  const { data } = await apiClient.get<SessionListResponse>('/journal/sessions', {
    params: filters,
  })
  return data
}

export async function getSession(id: string): Promise<SessionResponse> {
  const { data } = await apiClient.get<SessionResponse>(`/journal/sessions/${id}`)
  return data
}

export async function createSession(body: SessionCreate): Promise<SessionResponse> {
  const { data } = await apiClient.post<SessionResponse>('/journal/sessions', body)
  return data
}

export async function updateSession(id: string, body: SessionUpdate): Promise<SessionResponse> {
  const { data } = await apiClient.patch<SessionResponse>(`/journal/sessions/${id}`, body)
  return data
}

export async function deleteSession(id: string): Promise<void> {
  await apiClient.delete(`/journal/sessions/${id}`)
}
